src/Features/Core/Portable/MetadataAsSource/MetadataAsSourceHelpers.cs

```diff
@@ -96,28 +103,243 @@ public static async Task<Location> GetLocationInGeneratedSourceAsync(SymbolKey s
+        /// that handle, following any type forwards from the original symbol
+        /// </summary>
+        public static EntityHandle? FindEntityHandle(ISymbol symbol, string dllPath, IPdbSourceDocumentLogger? logger, out string resolvedDllPath)
+        {
+            resolvedDllPath = dllPath;
+
+            // We have an implementation assembly, but it might not be the right one. We need to follow any type forwards
+            // that there might be, to find the handle for symbol in the DLL specified by dllPath. Normally we could
+            // use symbol.MetadataToken to get the token, but that token could come from a reference assembly, so could
+            // be different. We need to load the DLL, and find the containing type, in order to find the entity handle
+            // for the symbol we want to find.
+            var containingTypeSymbol = symbol is INamedTypeSymbol typeSymbol ? typeSymbol : symbol.ContainingType;
+
+            // If we find any type forwards we'll assume they're in the same directory
+            var basePath = Path.GetDirectoryName(dllPath);
+            if (basePath is null)
+                return null;
+
+            try
+            {
+                while (File.Exists(resolvedDllPath))
+                {
+                    using var fileStream = File.OpenRead(resolvedDllPath);
+                    using var reader = new PEReader(fileStream);
+                    var md = reader.GetMetadataReader();
+
+                    // First, see if this type is exported to a different assembly
+                    var assemblyName = GetExportedTypeAssemblyName(md, containingTypeSymbol);
+                    if (assemblyName is not null)
+                    {
+                        resolvedDllPath = Path.Combine(basePath, $"{assemblyName}.dll");
+                        logger?.Log(FeaturesResources.Symbol_found_in_assembly_path_0, resolvedDllPath);
+                        continue;
+                    }
+
+                    // The type should be in this assembly, so lets find its handle
+                    var typeDefinition = GetTypeDefinition(md, containingTypeSymbol, out var typeDefHandle);
+                    if (typeDefinition is null)
+                        return null;
+
+                    // If we're looking for a type, we're already here
+                    if (symbol.Kind == SymbolKind.NamedType)
+                        return typeDefHandle;
+
+                    // Now we have the type, lets find the actual symbol
+                    return FindEntityHandle(md, typeDefinition.Value, symbol);
+                }
+            }
+            catch (Exception ex) when (IOUtilities.IsNormalIOException(ex))
+            {
+            }
+
+            return null;
+        }
+
+        private static string? GetExportedTypeAssemblyName(MetadataReader md, INamedTypeSymbol typeSymbol)
+        {
+            foreach (var eth in md.ExportedTypes)
```

Comment:
This is an annoyingly long search, as is the one to find the type, but I'm not sure if we should be just caching types and results as we find them, or building a map of the world as we go through things. Thoughts?

Considering the files on disk are presumably immutable, a map from ref dll + typename <-> impl dll + type definition handle probably makes sense to serialize to the .vs folder.

Reply:
This should be fast one the file is read from disk. I think the best caching would be such that avoids opening one file multiple times. Essentially caching all type-forwarders for each file we encounter.